File: Legacy/Leetcode/90SubsetsII.py
# Given a collection of integers that might contain duplicates, nums, return all possible subsets.
#
# Note: The solution set must not contain duplicate subsets.
#
# For example,
# If nums = [1,2,2], a solution is:
#
# [
#   [2],
#   [1],
#   [1,2,2],
#   [2,2],
#   [1,2],
#   []
# ]

import itertools

def subsetsWithDup(nums):
    length = len(nums)
    ans = []
    for i in range(length + 1):
        combs = itertools.combinations(nums, i)
        for comb in combs:
            comb = sorted(list(comb))
            if comb not in ans:
                ans.append(comb)
    return ans

File: C++/Backtracking/90. Subsets II.cpp
class Solution {
public:
    
    void backtrack(vector<int>& nums, vector<vector<int>>& res, vector<int>& comb, int start) {
        res.push_back(comb);
        for (int i = start; i < nums.size(); i++) {
            if (i != start && nums[i] == nums[i-1]) continue;
            comb.push_back(nums[i]);
            backtrack(nums, res, comb, i+1);
            comb.pop_back();
        }
    }
    
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> comb;
        sort(begin(nums), end(nums));
        backtrack(nums, res, comb, 0);
        return res;
    }
};